sycl/doc/SYCL2020-SpecializationConstants.md

```diff
@@ -0,0 +1,765 @@
+
+In order for runtime to access the right property, it need to compute the
+symbolic ID of a specialization constant based on user-provided inputs, such
+as non-type template argument passed to `set_specialization_constant` argument.
+DPC++ Headers section describes how symbolic IDs are generated and the same
+trick is used within `set_specialization_constant` method:
+```
+template<auto& SpecName>
+void set_specialization_constant(
+  typename std::remove_reference_t<decltype(SpecName)>::type value) {
+  const char *SymbolicID =
+#if __has_builint(__builtin_unique_stable_name)
+      __builtin_unique_stable_name(detail::specialization_id_name_generator<S>);
+#else
+      // without the builtin we can't get the symbolic ID of the constant
+      "";
+#endif
+  // remember the value of the specialization constant
+  SpecConstantValuesMap[SymbolicID] = value;
+}
+```
+
+The major downside of that approach is that it can't be used with any
+third-party host compiler, because it uses a specific built-in function to
+generate symbolic IDs of specialization constants. Good solution would be to
+employ integration header here, i.e. we could provide some class template
+specializations which will return symbolic IDs - the same approach as we use
+for communicating OpenCL kernel names from the compiler to the runtime.
+
+For the following user code:
+```
+specalization_id<int> id_int;
+// ...
+  [=](kernel_handler h) {
+    h.get_specialization_constant<id_int>();
+  }
+```
+
+The following integration header would be produced:
+```
+// fallback
+template<auto &S>
+class specialization_constant_info {
+  static const char *getName() { return ""; }
+};
+
+// forward declaration
+extern specialization_id<int> id_int;
+
+// specialization
+template<>
+class specialization_constant_info<id_int> {
+  static const char *getName() {
+    return "result of __builtin_unique_stable_name(detail::specialization_id_name_generator<id_int>) encoded here";
+  }
+};
+```
+
+And it would be used by DPC++ RT in the following way:
+```
+template<auto& SpecName>
+void set_specialization_constant(
+  typename std::remove_reference_t<decltype(SpecName)>::type value) {
+  const char *SymbolicID = specialiation_constant_info<SpecName>::getName();
+  // remember the value of the specialization constant
+  SpecConstantValuesMap[SymbolicID] = value;
+}
+```
+
+Such trick would allow use to compile host part of the app with any third-party
+compiler that supports C++17, but the problem here is that SYCL 2020 spec states
+the following:
+
+> Specialization constants must be declared using the `specialization_id` class,
+> and the declaration must be outside of kernel scope using static storage
+> duration. The declaration must be in either namespace scope or class scope.
+
+`class` scope `static` variables are not forward-declarable, which means that
+the approach with integration header is not available for us here.
```

Comment:
As an alternative, could we generate a new translation unit which is compiled and linked into the user's application?  With this approach, the new translation unit could be compiled with any C++17 host compiler, even a third party compiler.  To illustrate, assume the user's application had this:

```
specialization_id<int> id_int;

struct foo {
  static specalization_id<int> id_foo;
};
specialization_id<int> foo::id_foo;

// ...
  [=](kernel_handler h) {
    h.get_specialization_constant<id_int>();
    h.get_specialization_constant<foo::id_foo>();
  }
```

The generated translation unit could look like this:

```
template<auto &S>
const char *get_specialization_name();

extern specialization_id<int> id_int;
struct foo {
  static specalization_id<int> id_foo;
};

template<>
const char *get_specialization_name<id_int>() {
  return "result of __builtin_unique_stable_name(detail::specialization_id_name_generator<id_int>) encoded here";
}
template<>
const char *get_specialization_name<foo::id_foo>() {
  return "result of __builtin_unique_stable_name(detail::specialization_id_name_generator<foo::id_foo>) encoded here";
}
```

As you can see, this works even for class scope static variables.

Reply:
One more alternative I suggested in internal discussions. It does not need to generate a new source and relies on existing C++ symbol resolution mechanism in C++ compiler and linker, thanks to 2020 spec constant identification style based on static/global variables rather than types.

1) Similarly to what it does today, `sycl-post-link` tool creates a list of all the specialization constant (globals) used in the device code and assigns integer ID which will then serve as SPIR-V IDs
2) SYCL runtime is extended with a function `void register_spec_constant_map(map_type* map_addr)`
3) next, the post-link tool generates a host object – e.g. `sc.obj` – containing
    -  a static variable – a table whose rows contain 2 elements each: `[ <spec constant global address>, <its ID> ]`.
    -  a function in static constructor section. This function just calls register_spec_constant_map passing the table address to it.
  The object is generated using LLVM API similarly to how the clang-offload-wrapper tool creates a wrapper object - it first generates LLVM IR then native host object (ELF or COFF) is generated from it using llvm `llc` tool. The `<spec constant global address>` in the table is an external reference to the corresponding `specialization_id`  static variable.
4) Clang driver links the the `sc.obj` together with the application. Linkage will resolve the `<spec constant global address>` in the table.
    -  Theoretically, there can be more than one such objects linked with the same spec id var mapped to different IDs. In this case we return to the problem of device module identification which already exists and should be solved separately.
5) Upon startup, the app will register all the spec constants with SYCL runtime through `register_spec_constant_map`, now SYCL runtime is able to map `specialization_id` variable address to its SPIR-V id using the registered tables.

Reply:
Had a sync with @gmlueck about what approach from the two above - based on source generation (a) or .obj generation (b) - is better. My main concerns with (a) were:

-	Custom resolution scheme based on __builtin_unique_stable_name is used for variable resolution instead of std C++. String-based resolution for 1.2.1 was the only way, because spec constants were identified by types which don’t have linkage. In 2020 they are identified by static/global variables, which have linkage, so we can use std C++.
-	Extra C++ source -> obj compilation needs to happen, which adds overhead

During the sync, Greg suggested that `const char *get_specialization_name<...>()` in apporach a) is replaced by `int get_specialization_spirv_int_id<...>()` (generated by sycl-post-link), which returns the assigned spec constant ID w/o intermediate string id generation. In this case the first concern goes away, but the second still remains. Besides, since the source generation now must be done by post-link, the `specialization_id` variable names (`id_int` and `foo::id_foo` above) needs to be somehow propagated to the post-link tool, which may add complexity.

Overall, we are slightly more inclined to approach (b) – obj generation (e.g. it avoids the extra FE invocation at link time), but it is better for the feature owner to decide (may be based on factors which are not obvious at this point). @gmlueck, please confirm, or correct me here.

Reply:
Suggest that we preliminary take approach (b) then.

Reply:
Thanks, @kbobrovs for the summary of our meeting.  Yes, I also agree that option (b) is probably easier.  As you note, option (a) requires that we somehow save some *source code* in the object file so that the post-link tool can create the translation unit I describe.  This would be possible, but it seems awkward.  Option (a), instead, generates the translation unit directly from LLVM IR at post-link time.

Even if we do option (b), though, I think we should consider generating an object file (at post-link time) that contains the equivalent of the following C++:

```
template<auto &S>
int get_specialization_id();

extern specialization_id<int> id_int;
struct foo {
  static specalization_id<int> id_foo;
};

template<>
int get_specialization_id<id_int>() {
  return /* SPIR-V ID for "id_int" */;
}
template<>
int get_specialization_id<foo::id_foo>() {
  return /* SPIR-V ID for "foo::id_foo" */;
}
```

This seems like it might be simpler than generating the static table that Konst describes because it doesn't require registration of the map at start-up time, and it doesn't require any lookups in the map at runtime.

To illustrate, here is a simple prototype of the host code for `specialization_id` and the related `handler` member functions.  The prototype makes the simplifying assumption that specialization constants are always `int`, but this could be extended to allow arbitrary types using the design that @romanovvlad has already proposed.

```
#include <map>

namespace sycl {

class specialization_id {
 public:
  constexpr specialization_id(int i) : default_value(i) {}
 private:
  int default_value;
  friend class handler;
};


namespace detail {

// The post-link tool generates specializations of this function for each
// "specialization_id" variable in the application.
template<auto &SpecName>
int get_specialization_id();

} // namespace detail


class handler {
 public:
  template<auto& SpecName>
  void set_specialization_constant(int value) {
    spec_values[detail::get_specialization_id<SpecName>()] = value;
  }

  template<auto& SpecName>
  int get_specialization_constant() {
    int id = detail::get_specialization_id<SpecName>();
    auto it = spec_values.find(id);
    if (it == spec_values.end())
      return SpecName.default_value;
    return it->second;
  }

 private:
  std::map<int, int> spec_values;
};

} // namespace sycl
```

As @romanovvlad points out, you need extra information about each `specialization_id` in order to support arbitrary types because larger types may require more than one specialization ID.  Rather than storing this information in the device image property set, though, it might be easier to extend the `get_specialization_id()` function that the post-link tool generates.  Instead of returning a simple ID, it could return a struct with all the information you need (array of IDs, etc.)

Reply:
@gmlueck, generating `get_specialization_id<foo::id_foo>()` and bypassing the need in table lookup during host-side spec constant resolution seems reasonable to me. If we ever need the table, we may additionally generate it.
Creating a mangled function name in the IR for `get_specialization_id<foo::id_foo>` might be little tricky, need to use some C++ mangler.

Reply:
> @gmlueck, generating `get_specialization_id<foo::id_foo>()` and bypassing the need in table lookup during host-side spec constant resolution seems reasonable to me. If we ever need the table, we may additionally generate it.

The approach seems reasonable to me as well, but 

> Creating a mangled function name in the IR for `get_specialization_id<foo::id_foo>` might be little tricky, need to use some C++ mangler.

I'm afraid this might be quite tricky at LLVM IR level, because we should recreate proper mangling for non-type template parameters, but together with @romanovvlad we have brainstormed on another idea: what if we do the same at DPC++ FE level? We could generate `get_specialization_id<foo::id_foo>` definition in already mangled form to avoid forward-declaring `specialization_id`s (which can't be forward-declared in general case, due to the spec wording, i.e. non-`inline` `constexpr` or just `static` `specialization_id`s are allowed). Such definition would return a unique string which will be used as interface between the compiler and the runtime like it is used right now (but we probably need something different from `__builtin_unque_stable_name`, because we need to get different strings for variables of the same type with the same name defined in different translation units).

Pseudo-code example:
```
// user code:
specialization_id<int> id_const;
// ...
cgh.set_specialization_constant<id_const>(42);
// ...
kh.get_specialization_constant<id_const>();

// implementation:
template<auto &SpecName>
kernel_handler::get_specialization_constant() {
#ifdef __SYCL_DEVICE_ONLY__
  // on device side we use compiler built-in directly (as we don't have integration header available)
  const char *SymbolicID = __get_symbolic_spec_id();
  return __sycl_getSYCL2020SpecializationConstant(SymbolicID, ...);
#endif 

template<auto &SpecName>
kernel_handler::set_specialization_constant() {
  // on host we use special mapping function
  const char *SymbolicID = detail::get_symbolic_spec_id<SpecName>();
  // lookup device image properties as in the current implementation
}

// that mapping function is provided through integration header:
extern "C" static const char *_ZN6detail20get_symbolic_spec_idIL_ZL8id_constEEEPKcv() {
  return "result of __get_symbolic_spec_id() built-in for id_const here";
}
// _ZN6detail20get_symbolic_spec_idIL_ZL8id_constEEEPKcv is a mangling of detail::get_symbolic_spec_id<id_const>
```

Pros of this approach:
- I assume that it is easier to work with mangling at AST level than at LLVM IR level (tagging @erichkeane here to comment)
- we can re-use existing device image properties mechanism
Cons:
- Seems hacky
- Technically we do not account for `specialization_id` linkage and only rely on some generated names

Reply:
New design approach was chosen and documented in a0832d5d83ba1411522a02dc5c8cc5b95aca1fa1, marking this conversation as resolved